subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -53,118 +36,21 @@ public DefaultPropertyWalker(PropertyMetadataStore propertyMetadataStore) {
 
     @Override
     public void visitProperties(PropertySpecFactory specFactory, PropertyVisitor visitor, Object bean) {
-        Queue<NestedBeanContext> queue = new ArrayDeque<NestedBeanContext>();
-        queue.add(new NestedBeanContext(BeanNode.create(null, bean), queue, null));
+        Queue<BeanNodeContext> queue = new ArrayDeque<BeanNodeContext>();
+        queue.add(new BeanNodeContext(BeanNode.create(null, bean, propertyMetadataStore), queue, null));
```

Comment:
Those two `null` parameters make things harder to read than I'd like. :/

Reply:
Add an own class for the root.